module/hotstuff.go

```diff
@@ -21,32 +27,49 @@ type HotStuff interface {
-// about finalization of blocks. It consumes block proposals broadcasted
-// by the consensus node, verifies the block header and locally evaluates
-// the finalization rules.
+// In contrast to an active HotStuff participant, the HotStuffFollower does not validate
+// block payloads. This greatly reduces the amount of CPU and memory that it consumes.
+// Essentially, the consensus participants exhaustively verify the entire block including
+// the payload and only vote for the block if it is valid. The consensus committee
+// aggregates votes from a supermajority of participants to a Quorum Certificate [QC].
+// Thereby, it is guaranteed that only valid blocks get certified (receive a QC).
+// By only consuming certified blocks, the HotStuffFollower can be sure of their
+// correctness and omit the heavy payload verification.
+// There is no disbenefit for nodes to wait for a QC (included in child blocks), because
+// all nodes other than consensus generally require the Source Of Randomness included in
+// QCs to process the block in the first place.
+//
+// The central purpose of the HotStuffFollower is to informs other components within the
+// node about finalization of blocks.
 //
 // Notes:
-//   - HotStuffFollower does not handle disconnected blocks. Each block's parent must
-//     have been previously processed by the HotStuffFollower.
 //   - HotStuffFollower internally prunes blocks below the last finalized view.
-//     When receiving a block proposal, it might not have the proposal's parent anymore.
-//     Nevertheless, HotStuffFollower needs the parent's view, which must be supplied
-//     in addition to the proposal.
+//   - HotStuffFollower does not handle disconnected blocks. For each input block,
+//     we require that the parent was previously added (unless the parent's view
+//     is _below_ the latest finalized view).
 type HotStuffFollower interface {
 	ReadyDoneAware
 	Startable
 
-	// SubmitProposal feeds a new block proposal into the HotStuffFollower.
-	// This method blocks until the proposal is accepted to the event queue.
+	// AddCertifiedBlock appends the given certified block to the tree of pending
+	// blocks and updates the latest finalized block (if finalization progressed).
+	// Unless the parent is below the pruning threshold (latest finalized view), we
+	// require that the parent has previously been added.
 	//
-	// Block proposals must be submitted in order, i.e. a proposal's parent must
-	// have been previously processed by the HotStuffFollower.
-	SubmitProposal(proposal *model.Proposal)
+	// Notes:
+	//  - Under normal operations, this method is non-blocking. The follower internally
+	//    queues incoming blocks and processes them in its own worker routine. However,
+	//    when the inbound queue is, we block until there is space in the queue. This
+	//    behaviours is intentional, because we cannot drop blocks (otherwise, we would
```

Comment:
```suggestion
	//    when the inbound queue is full, we block until there is space in the queue. This
	//    behaviour is intentional, because we cannot drop blocks (otherwise, we would
```

Reply:
applied in https://github.com/onflow/flow-go/pull/4234/commits/3fa1776152863588863cd7d6c98ab137de5d2e3f